### PR TITLE
Request delegation

### PR DESCRIPTION
**Pull request by adrianschroeter**
[api] support request delegation to package origin project.
    
    Adds an option to let the server delegate the request for
    a target project to the project where the packages is actual
    living.
    
    This works when source has no package link, eg from submitting
    from a base distro project to another one.
    
    This is needed for Jump submissions in first place (jsc#OBS-59)

### CODE REVIEW

**Comment by adrianschroeter** [comment]
@hennevogel it would be nice if someone from frontend team can look at the failures. (I do not understand how the feature failure can be related to my change and I do not know how to make the new default attrib_type visible for the rspec testsuite)

It is relative time critical since this blocks the submission to jump distro.

**Comment by vpereira** [comment]
@adrianschroeter please rebase your branch with master. The issue is fixed in master. We have some flickering tests, in mobile resolution, which are disabled on master.

File: src/api/app/models/bs_request.rb [hunk 1]
```diff
@@ -293,6 +293,10 @@ def set_ignore_build_state
     @ignore_build_state = true
   end
 
+  def set_ignore_delegate
+    @ignore_delegate = true
```
Comment (vpereira) [inline src/api/app/models/bs_request.rb:297]: attr_acessor, no?

File: src/api/app/controllers/request_controller.rb [hunk 2]
```diff
@@ -146,6 +146,7 @@ def request_create
       @req = BsRequest.new_from_xml(request.raw_post.to_s)
       authorize @req, :create?
       @req.set_add_revision       if params[:addrevision].present?
+      @req.set_ignore_delegate    if params[:ignore_delegate].present?
```
Comment (vpereira) [inline src/api/app/controllers/request_controller.rb:149]: please use an attr_acessor for that..

File: src/api/app/models/bs_request.rb [hunk 3]
```diff
@@ -1031,7 +1035,7 @@ def expand_targets
     oldactions = []
 
     bs_request_actions.each do |action|
-      na, ppl = action.expand_targets(!@ignore_build_state.nil?)
+      na, ppl = action.expand_targets(!@ignore_build_state.nil?, !@ignore_delegate.nil?)
```
Comment (vpereira) [inline src/api/app/models/bs_request.rb:1038]: `@ignore_delegate.present?`

File: src/api/app/models/bs_request_action.rb [hunk 4]
```diff
@@ -663,9 +663,34 @@ def check_action_permission!(skip_source = nil)
     check_permissions!
   end
 
-  def expand_targets(ignore_build_state)
+  def expand_targets(ignore_build_state, ignore_delegate)
     # expand target_package
 
+    # target delegation handling
+    if action_type == :submit
```
Comment (vpereira) [inline src/api/app/models/bs_request_action.rb:670]: we check two times if the action type is submit. Look line 694:

```if action_type.in?([:submit, :maintenance_incident])```

**Comment by adrianschroeter** [comment]
On Donnerstag, 28. Mai 2020, 09:37:21 CEST wrote Victor Pereira:
> @adrianschroeter please rebase your branch with master. The issue is fixed in master. We have some flickering tests, in mobile resolution, which are disabled on master.

done, but I doubt that the missing attribute issue in rspec will be solved by that?

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

File: src/api/db/data/20200627135426_request_delegate_attribute.rb [hunk 5]
```diff
@@ -0,0 +1,20 @@
+require_relative '../attribute_descriptions'
+
+class RequestDelegateAttribute < ActiveRecord::Migration[4.2]
```
Comment (vpereira) [inline src/api/db/data/20200627135426_request_delegate_attribute.rb:3]: RequestDelegateAttribute < ActiveRecord::Migration[6.0]

**Comment by coolo** [comment]
Am I the only one with no clue what this feature is supposed to do?

**Comment by adrianschroeter** [comment]
On Donnerstag, 28. Mai 2020, 09:46:48 CEST wrote Stephan Kulow:
> Am I the only one with no clue what this feature is supposed to do? 

It delegate the request target to project where package is living.
basically replacing the given target_project based on project policy.
See the jira...

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by vpereira** [comment]
> done, but I doubt that the missing attribute issue in rspec will be solved by that?

@adrianschroeter  well specs broken by your code couldn't be fixed on master. The feature specs are fixed.

**Comment by adrianschroeter** [comment]
On Donnerstag, 28. Mai 2020, 10:12:11 CEST wrote Victor Pereira:
> > done, but I doubt that the missing attribute issue in rspec will be solved by that?
> 
> @adrianschroeter  well specs broken by your code couldn't be fixed on master. The feature specs are fixed. 

right, and how could learn rspec about the new default attribute?

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by adrianschroeter** [comment]
On Donnerstag, 28. Mai 2020, 09:45:37 CEST wrote Victor Pereira:
> @vpereira requested changes on this pull request.
> 
> 
> 
> > @@ -293,6 +293,10 @@ def set_ignore_build_state
>      @ignore_build_state = true
>    end
>  
> +  def set_ignore_delegate
> +    @ignore_delegate = true
> 
> attr_acessor, no?
> 
> > @@ -146,6 +146,7 @@ def request_create
>        @req = BsRequest.new_from_xml(request.raw_post.to_s)
>        authorize @req, :create?
>        @req.set_add_revision       if params[:addrevision].present?
> +      @req.set_ignore_delegate    if params[:ignore_delegate].present?
> 
> please use an attr_acessor for that.. 

well, tried to be consistent to existent code, but we can change that.

> > @@ -1031,7 +1035,7 @@ def expand_targets
>      oldactions = []
>  
>      bs_request_actions.each do |action|
> -      na, ppl = action.expand_targets(!@ignore_build_state.nil?)
> +      na, ppl = action.expand_targets(!@ignore_build_state.nil?, !@ignore_delegate.nil?)
> 
> `@ignore_delegate.present?`
> 
> >      # expand target_package
>  
> +    # target delegation handling
> +    if action_type == :submit
> 
> we check two times if the action type is submit. Look line 694:

yes, but this code must not be used in maintenance_incident case.


-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

File: src/api/db/data/20200627135426_request_delegate_attribute.rb (hunk 5 above)
Comment (adrianschroeter) [reply src/api/db/data/20200627135426_request_delegate_attribute.rb]: changed

**Comment by adrianschroeter** [comment]
so, can we do the attr_acessor usage in another step and cover not only this, but also the other two existing cases there for consitency?

**Comment by vpereira** [comment]
> right, and how could learn rspec about the new default attribute?

As usual, you can either create a Factory or in the rspec associate it to the object. You can check the `project_spec.rb` how we do for example with other `AttribType`

**Comment by vpereira** [comment]
> so, can we do the attr_acessor usage in another step and cover not only this, but also the other two existing cases there for consitency?

not sure if understood it, but do you wan't to fix it now? Why do you want to create a custom setter, if ruby offers you it for free? https://apidock.com/ruby/Module/attr_accessor

**Comment by adrianschroeter** [comment]
On Donnerstag, 28. Mai 2020, 11:58:16 CEST wrote Victor Pereira:
> > right, and how could learn rspec about the new default attribute?
> 
> As usual, you can either create a Factory or in the rspec associate it to the object. You can check the `project_spec.rb` how we do for example with other `AttribType` 

project_spec.rb is also just using the default attributes.

I suppose some vcr cassettes need  to be re-recorded, but I can not do that.

Could some take this over please

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by adrianschroeter** [comment]
On Donnerstag, 28. Mai 2020, 12:00:59 CEST wrote Victor Pereira:
> > so, can we do the attr_acessor usage in another step and cover not only this, but also the other two existing cases there for consitency?
> 
> not sure if understood it, but do you wan't to fix it now? Why do you want to create a custom setter, if ruby offers you it for free? https://apidock.com/ruby/Module/attr_accessor

because I have just added another option beside others. It was not my idea to style
it that way for the other options, but it was some refactoring before (and before
that it was just a variable).

however, I do not want to introduce an inconsitency now but stick to the current 
practice. I do not mind if you want to change it in another refactoring sessions,
but this is out of the scope of this submission.

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Comment by vpereira** [comment]
> however, I do not want to introduce an inconsitency now but stick to the current practice. I do not mind if you want to change it in another refactoring sessions, but this is out of the scope of this submission.

ok I assigned @hennevogel  and I will drop my review

File: src/api/app/controllers/request_controller.rb (hunk 2 above)
Comment (hennevogel) [reply src/api/app/controllers/request_controller.rb:149]: A refactoring for another day...

File: src/api/app/models/bs_request_action.rb (hunk 4 above)
Comment (hennevogel) [reply src/api/app/models/bs_request_action.rb]: Yeah but this is just for submit requests, not for maintenance incident requests.

**Comment by hennevogel** [comment]
I ignored code climate, it's impossible to reduce the complexity of the method without making this more unreadable. We tried for a while...

**Comment by adrianschroeter** [comment]
thanks!

File: src/api/db/data/20200627135426_request_delegate_attribute.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+require_relative '../attribute_descriptions'
```
Comment (DavidKang) [inline src/api/db/data/20200627135426_request_delegate_attribute.rb:1]: @adrianschroeter and @hennevogel, the date of this migration '20200627135426_request_delegate_attribute.rb' is on future for a reason?

**Comment by adrianschroeter** [comment]
On Montag, 1. Juni 2020, 15:46:48 CEST wrote David Kang:
> @DavidKang commented on this pull request.
> 
> 
> 
> > @@ -0,0 +1,20 @@
> +require_relative '../attribute_descriptions'
> 
> @adrianschroeter and @hennevogel, the date of this migration '20200627135426_request_delegate_attribute.rb' is on future for a reason? 

no ... just fat fingers (shouldn't be a problem though)

-- 

Adrian Schroeter
email: adrian@suse.de

SUSE Linux GmbH, GF: Felix Imendörffer, Jane Smithard, Graham Norton, HRB 21284 (AG Nürnberg) 
 
Maxfeldstraße 5                         
90409 Nürnberg 
Germany

**Pull request merged**